### PR TITLE
Display vertebral levels when provided with `-perslice 1`

### LINKED ISSUE #3113: Display vertebral level when using -perslice 1 option
When running `sct_process_segmentation` (or `sct_extract_metric`) with option `perslice 1`, each row corresponds to a slice. However, when doing so, the column "VertLevel" is empty:

Filename | Slice (I->S) | VertLevel | MEAN(area)
-- | -- | -- | --
/Users/julien/temp/20201217_inspired/results6/data_processed/HC02/bl/cord/processing_sct/t2_tra_seg.nii.gz | 0 |   | 70.41316501066880
/Users/julien/temp/20201217_inspired/results6/data_processed/HC02/bl/cord/processing_sct/t2_tra_seg.nii.gz | 1 |   | 76.68867922117070
/Users/julien/temp/20201217_inspired/results6/data_processed/HC02/bl/cord/processing_sct/t2_tra_seg.nii.gz | 2 |   | 76.84621033344560
/Users/julien/temp/20201217_inspired/results6/data_processed/HC02/bl/cord/processing_sct/t2_tra_seg.nii.gz | 3 |   | 81.27146085321930
/Users/julien/temp/20201217_inspired/results6/data_processed/HC02/bl/cord/processing_sct/t2_tra_seg.nii.gz | 4 |   | 82.83952355053640

It shouldn't be empty, because we _know_ the information about vertebral level, if user input vertebral labeling file, as was done here, with the flag `-vertfile`. 

So as a feature, we should fill this column "VertLevel" when the vertebral level for each slice is known.

### PR DESCRIPTION
**Pull request by sandrinebedard**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [x] I've linked relevant issues in the PR body
- [x] I've applied [the relevant labels](https://intranet.neuro.polymtl.ca/geek-tips/contributing#pr-labels-a-href-pr-labels-id-pr-labels-a) to this PR
- [ ] I've applied a [release milestone](https://github.com/spinalcordtoolbox/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [x] I've assigned a reviewer

#### PR contents

- [x] I've consulted [SCT's internal developer documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [ ] I've updated the [relevant documentation](https://github.com/spinalcordtoolbox/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
This PR adresses #3113. In `sct_process_segmentation` when specifying `-vertfile` and `-perslice 1`, it now specifies the vertebral levels (when they are provided). 

## Linked issues
Fixes #3113 (and also its duplication #3976)

### CODE REVIEW

File: spinalcordtoolbox/aggregate_slicewise.py [hunk 1]
```diff
@@ -324,6 +329,8 @@ def aggregate_per_slice_or_level(metric, mask=None, slices=[], levels=[], distan
             agg_metric[slicegroup]['VertLevel'] = None
         else:
             agg_metric[slicegroup]['VertLevel'] = vertgroups[slicegroups.index(slicegroup)]
+            if agg_metric[slicegroup]['VertLevel'][0] is None:
+                agg_metric[slicegroup]['VertLevel'] = None
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/aggregate_slicewise.py:333]: > I'm trying to understand commit https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4004/commits/0c21b168b275d01caa83b152d0b7018b7445c58e. It seems like it's made up of three parts?

3. **?**

    My apologies, but on first read, I don't quite understand this change. What's the purpose of this change?

**Review by joshuacwnewton — COMMENTED** [review]
Thank you for taking on this task, and opening up a PR so quickly! :hearts: 

I found this PR a little difficult to review, though. So, I have some [PR ettiquette-related](https://medium.com/google-developer-experts/how-to-pull-request-d75ac81449a5#e07e) requests for the future, if possible?

- Could you add [extended descriptions](https://stackoverflow.com/questions/9562304/git-github-commit-with-extended-message-description) to your commits, when applicable? 
  
  They're really helpful for explaining the reasons for each change. Without explanations, a reviewer either has to has to guess the motivations for each commit, or they have to take the time ask for explanations as part of their review. This slows down the review process, and makes a PR harder to review.
  
  (Obviously sometimes the changes are pretty self-explanatory? But in this case, the motivations for each commit weren't clear to me.)
- Could you break down commits into smaller changes? 

  (I had some trouble understanding https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/0c21b168b275d01caa83b152d0b7018b7445c58e, since it has multiple changes that are seemingly unrelated.)

Thank you again!

File: spinalcordtoolbox/scripts/sct_process_segmentation.py [hunk 2]
```diff
@@ -372,7 +371,7 @@ def main(argv: Sequence[str]):
 
     file_out = os.path.abspath(arguments.o)
     append = bool(arguments.append)
-    if arguments.vert is not None:
+    if arguments.vertfile is not None:
```
[the remark this answers is not in the document]
Comment (sandrinebedard) [reply spinalcordtoolbox/scripts/sct_process_segmentation.py:374]: The case we want to adress here is if we provide vertebral labels (`vertfile`) without having to specify `vert` which corresponds to the levels we want to include. Before, since the condition was on `arguments.vert`, it puts directly the vertfile to None, even if it was specified by the user.

File: spinalcordtoolbox/scripts/sct_process_segmentation.py [hunk 3]
```diff
@@ -153,7 +153,6 @@ def get_parser():
     optional.add_argument(
         '-vertfile',
         metavar=Metavar.str,
-        default=os.path.join('.', 'label', 'template', 'PAM50_levels.nii.gz'),
```
[the remark this answers is not in the document]
Comment (sandrinebedard) [reply spinalcordtoolbox/scripts/sct_process_segmentation.py:156]: `arguments.vertfile` will never be None if it is the template, conlfict with https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/3eaf44e1ab1211e0b86ef38137d2df16e14ffb45, 
I was curious about the motivation for this. Would be interesting to dicuss, I wasn't sure if it was the best idea to remove.

File: spinalcordtoolbox/aggregate_slicewise.py (hunk 1 above)
Comment (sandrinebedard) [reply spinalcordtoolbox/aggregate_slicewise.py:333]: Here, if there was no level corresponding to the slice, the tuple gives : `(None,)` which means that is the .csv file, it was written `None` instead of leaving it empty. This is just to put None instead of (None,)

File: spinalcordtoolbox/aggregate_slicewise.py [hunk 4]
```diff
@@ -258,6 +258,8 @@ def aggregate_per_slice_or_level(metric, mask=None, slices=[], levels=[], distan
             raise ValueError(f"Shape mismatch between vertfile [{vert_level_slices}] and metric [{metric_slices}]). "
                              f"Please verify that your vertfile has the same number of slices as your input image, "
                              f"and that your metric is RPI/LPI oriented.")
+        # Fetch fname_vert_level
+        im_vert_level = Image(fname_vert_level).change_orientation('RPI')
```
[the remark this answers is not in the document]
Comment (sandrinebedard) [reply spinalcordtoolbox/aggregate_slicewise.py:262]: moving it up was necessary since `im_vert_level` was only in the if statement ` if levels` which it dosn't enter in the case we want to address here. Instead of repeating it in the if statement for perslice

File: spinalcordtoolbox/scripts/sct_process_segmentation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_process_segmentation.py:156]: > I was curious about the motivation for this.

I think the original idea behind having this default was to allow users to specify just `-vert` without `-vertfile`. In that case, the script will assume the presence of the `PAM50_levels.nii.gz` file in the working directory, generated by `sct_warp_template`.

By removing this default, users will have to manually specify `-vertfile` when using `-vert`. This is a little extra work, and could break some old scripts, so I would want to see if there's a way to make https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/3eaf44e1ab1211e0b86ef38137d2df16e14ffb45 work without removing the default. 

I'm still thinking about this though... :thinking:

File: spinalcordtoolbox/scripts/sct_process_segmentation.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_process_segmentation.py:374]: Ahhh, I see. That makes sense.

Question: Do we need to have a conditional at all? If we just wrote:

```python
levels = arguments.vert
fname_vert_level = arguments.vertfile
```

And we made it  so that `arguments.vert` is `[]` by default (we would just have to add `default=''` to `arguments.vert`)...

Then the behavior would be identical, wouldn't it?

File: spinalcordtoolbox/aggregate_slicewise.py (hunk 1 above)
Comment (joshuacwnewton) [reply spinalcordtoolbox/aggregate_slicewise.py:332]: Ahhh, okay. So basically:

- `(None,)` --> "None" cell
- `None` --> Empty cell

And we prefer an empty cell here?

Comment (sandrinebedard) [reply spinalcordtoolbox/aggregate_slicewise.py:332]: yes, to be consistent with the other cells in the .csv file. If there is no segmentation for a particular slice, the cell is left empty when no metric are computed

File: spinalcordtoolbox/scripts/sct_process_segmentation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sandrinebedard) [reply spinalcordtoolbox/scripts/sct_process_segmentation.py:156]: great, I'll try that!

**Comment by joshuacwnewton** [comment]
I've made some changes to implement the suggestions from the [earlier review](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4004#pullrequestreview-1263934291).

To test these changes, I followed the instructions from the [Registration to Template tutorial](https://spinalcordtoolbox.com/user_section/tutorials/registration-to-template.html) to register an image to the PAM50 template. Then, I tried computing `-perslice` metrics on the segmentation, with and without the default `-vertfile` being present.

<details>
<summary> Before running <code>sct_warp_template</code> (warning is present, no vertlevel info) </summary>

```console
(sb/3113-display-vert-level) joshua@tadpole:~/repos/spinalcordtoolbox/data/sct_example_data/t2$ sct_process_segmentation -i t2_seg.nii.gz -z 30:35 -perslice 1 -o csa_perslice.csv

--
Spinal Cord Toolbox (git-sb/3113-display-vert-level-9dd71dc98a637e5aa2175e8149631e43e40265da)

sct_process_segmentation -i t2_seg.nii.gz -z 30:35 -perslice 1 -o csa_perslice.csv
--

Vertebral level file ./label/template/PAM50_levels.nii.gz does not exist. Vert level information will not be displayed. To use vertebral level information, you may need to run `sct_warp_template` to generate the appropriate level file in your working directory.
Compute shape analysis: 100%|##############| 240/240 [00:01<00:00, 194.14iter/s]

Done! To view results, type:
xdg-open /home/joshua/repos/spinalcordtoolbox/data/sct_example_data/t2/csa_perslice.csv
```

image

</details>

<details>
<summary> After running <code>sct_warp_template</code> (no warning, vertlevel info is present) </summary>

```console
(sb/3113-display-vert-level) joshua@tadpole:~/repos/spinalcordtoolbox/data/sct_example_data/t2$ sct_process_segmentation -i t2_seg.nii.gz -z 30:35 -perslice 1 -o csa_perslice.csv

--
Spinal Cord Toolbox (git-sb/3113-display-vert-level-9dd71dc98a637e5aa2175e8149631e43e40265da)

sct_process_segmentation -i t2_seg.nii.gz -z 30:35 -perslice 1 -o csa_perslice.csv
--

Compute shape analysis: 100%|##############| 240/240 [00:01<00:00, 125.36iter/s]

Done! To view results, type:
xdg-open /home/joshua/repos/spinalcordtoolbox/data/sct_example_data/t2/csa_perslice.csv
```

image

</details>

File: spinalcordtoolbox/scripts/sct_process_segmentation.py [hunk 5]
```diff
@@ -372,12 +373,15 @@ def main(argv: Sequence[str]):
...
-        fname_vert_level = None
+    levels = arguments.vert
+    fname_vert_level = arguments.vertfile
+    if not os.path.isfile(fname_vert_level):
+        logger.warning(f"Vertebral level file {fname_vert_level} does not exist. Vert level information will "
+                       f"not be displayed. To use vertebral level information, you may need to run "
+                       f"`sct_warp_template` to generate the appropriate level file in your working directory.")
+        fname_vert_level = None  # Discard the default '-vertfile', so that we don't attempt to find vertebral levels
+        if levels:
+            raise FileNotFoundError("The vertebral level file must exist to use `-vert` to group by vertebral level.")
```
Comment (joshuacwnewton) [inline spinalcordtoolbox/scripts/sct_process_segmentation.py:384]: Just to provide a little more context for why this change was made:

* The `-vertfile` argument has a default value of `./label/template/PAM50_levels.nii.gz`. (This corresponds to a path that gets created when the user has run `sct_warp_template`. Otherwise, this file will probably not exist.)
* Before this PR, to get around the issue of the default filepath potentially being nonexistent, we made the assumption that if the user has manually specified the related option, `-vert`, then they are signalling intent, and will have probably run `sct_warp_template` beforehand.
* So, `-vertfile` was only used if if `-vert` was also passed. (If `-vert` was not specified, then `-vertfile` would be overridden by `None`.) 
* However, this PR requires us to access the value of `-vertfile` even in cases where `-vert` isn't specified.
* This means we need a new way (apart from checking `-vert`) to ensure that we're not trying to access a nonexistent file.
* So, the new method directly checks for the existence of `-vertfile`, and sets it to `None` (and provides helpful warnings/errors) in the case where it doesn't exist.
